internal/backend/workloadproxy/reconciler.go

```diff
@@ -0,0 +1,208 @@
+	hostPort := registry.hostPortForAlias(cluster, alias)
+
+	if upstreamsChanged {
+		if lbSts == nil { // no LB yet, create and start it
+			tcpLB := &loadbalancer.TCP{
+				Proxy: tcpproxy.Proxy{
+					ListenFunc: registry.connProvider.Listen,
+				},
+				DialTimeout:    1 * time.Second,
+				TCPUserTimeout: 5 * time.Second,
+			}
+
+			if err := tcpLB.AddRoute(hostPort, upstreamAddresses); err != nil {
+				return fmt.Errorf("failed to add route for %q/%q: %w", cluster, alias, err)
+			}
+
+			if err := tcpLB.Start(); err != nil {
+				return fmt.Errorf("failed to start load balancer for %q/%q: %w", cluster, alias, err)
+			}
+
+			lbSts = &lbStatus{
+				lb: tcpLB,
+			}
+		} else { // there is an existing LB, update it
+			if err := lbSts.lb.ReconcileRoute(hostPort, upstreamAddresses); err != nil {
+				return fmt.Errorf("failed to reconcile route for %q/%q: %w", cluster, alias, err)
+			}
+		}
+	}
+
+	lbSts.upstreamAddresses = upstreamAddresses
+	registry.aliasToCluster[alias] = cluster
+
+	aliasToLB := registry.clusterToAliasToLBStatus[cluster]
+	if aliasToLB == nil {
+		aliasToLB = map[string]*lbStatus{}
+
+		registry.clusterToAliasToLBStatus[cluster] = aliasToLB
+	}
+
+	aliasToLB[alias] = lbSts
+
+	return nil
+}
+
+// GetProxy returns a proxy for the exposed service, targeting the load balancer for the given alias.
+func (registry *Reconciler) GetProxy(alias string) (http.Handler, resource.ID, error) {
+	registry.mu.Lock()
+	defer registry.mu.Unlock()
+
+	clusterID, ok := registry.aliasToCluster[alias]
+	if !ok {
+		return nil, "", nil
+	}
+
+	lbSts := registry.clusterToAliasToLBStatus[clusterID][alias]
+	if lbSts == nil || lbSts.lb == nil {
+		return nil, "", nil
+	}
+
+	hostPort := registry.hostPortForAlias(clusterID, alias)
+
+	targetURL := &url.URL{
+		Scheme: "http",
+		Host:   hostPort,
+	}
+
+	proxy := httputil.NewSingleHostReverseProxy(targetURL)
+	proxy.Transport = &http.Transport{
+		DialContext: registry.connProvider.DialContext,
+	}
+
+	return proxy, clusterID, nil
+}
+
+func (registry *Reconciler) removeLB(cluster resource.ID, alias string) error {
+	registry.logger.Log(registry.logLevel, "remove LB", zap.String("cluster", cluster), zap.String("alias", alias))
+
+	aliasToLB := registry.clusterToAliasToLBStatus[cluster]
+	lbSts := aliasToLB[alias]
+
+	if lbSts != nil && lbSts.lb != nil {
+		if err := lbSts.lb.Close(); err != nil {
+			return fmt.Errorf("failed to close load balancer: %w", err)
+		}
+	}
+
+	delete(aliasToLB, alias)
+	delete(registry.aliasToCluster, alias)
+
+	if len(aliasToLB) == 0 {
+		delete(registry.clusterToAliasToLBStatus, cluster)
+	}
+
+	return nil
+}
+
+// hostPortForAlias returns a unique IP:port for the given cluster and alias.
+//
+// The value is arbitrary, as it uses in-memory transport, never reached via the network.
+func (registry *Reconciler) hostPortForAlias(clusterID resource.ID, alias string) string {
+	return fmt.Sprintf("%s_%s:4242", clusterID, alias)
```

Comment:
What stops us from running lb on the exposed service port directly?
Why do we need memu layer?

Reply:
I thought about that, but then we would need to allocate a distinct port or a loopback address for each and every exposed svc. I initially thought something like:
```
exposed svc 1 -> 127.0.0.1:40001
exposed svc 2 -> 127.0.0.1:40002
```
or
```
exposed svc 1 -> 127.0.40.1:40000
exposed svc 2 -> 127.0.40.2:40000
```

and so on.

But I didn't like that solution much, as when Omni receives the request for the exposed service, the data is already in-memory. It just needs to hit one of the healthy upstreams. It does not need to leave the process at all, just to come back to it to be forwarded to the upstream. So this felt like a more efficient and neater solution to me.